src/rtctools_heat_network/workflows/grow_workflow.py

```diff
@@ -523,18 +539,58 @@ def run_end_scenario_sizing(end_scenario_problem_class, staged_pipe_optimization
     if staged_pipe_optimization:
         solution = run_optimization_problem(end_scenario_problem_class, stage=1)
         results = solution.extract_results()
+        parameters = solution.parameters(0)
 
         # We give bounds for stage 2 by allowing one DN sizes larger than what was found in the
         # stage 1 optimization.
         pc_map = solution._HeatMixin__pipe_topo_pipe_class_map
         for pipe_classes in pc_map.values():
             v_prev = 0.0
+            first_pipe_class = True
             for var_name in pipe_classes.values():
                 v = results[var_name][0]
-                boolean_bounds[var_name] = (0.0, abs(v))
-                if v_prev == 1.0:
+                if first_pipe_class and abs(v) == 1.0:
+                    boolean_bounds[var_name] = (abs(v), abs(v))
+                elif abs(v) == 1.0:
+                    boolean_bounds[var_name] = (0.0, abs(v))
+                elif v_prev == 1.0:
                     boolean_bounds[var_name] = (0.0, 1.0)
+                else:
+                    boolean_bounds[var_name] = (abs(v), abs(v))
                 v_prev = v
+                first_pipe_class = False
+
+        for asset in [
+            *solution.heat_network_components.get("source", []),
+            *solution.heat_network_components.get("buffer", []),
+        ]:
+            var_name = f"{asset}_aggregation_count"
+            r = results[var_name][0]
+            if round(r) >= 1:
+                boolean_bounds[var_name] = (r, r)
+
+        t = solution.times()
+        from rtctools.optimization.timeseries import Timeseries
+
+        for p in solution.heat_network_components.get("pipe", []):
+            if p in solution.hot_pipes and parameters[f"{p}.area"] > 0.0:
+                lb = []
+                ub = []
+                for i in range(len(t)):
+                    r = results[f"{p}__flow_direct_var"][i]
+                    lb.append(
+                        r if abs(results[f"{p}.Q"][i] / parameters[f"{p}.area"]) > 1.0e-2 else 0
+                    )
+                    ub.append(
+                        r if abs(results[f"{p}.Q"][i] / parameters[f"{p}.area"]) > 1.0e-2 else 1
+                    )
+
+                boolean_bounds[f"{p}__flow_direct_var"] = (Timeseries(t, lb), Timeseries(t, ub))
+                try:
+                    r = results[f"{p}__is_disconnected"]
+                    boolean_bounds[f"{p}__is_disconnected"] = (Timeseries(t, r), Timeseries(t, r))
```

Comment:
not yet sure..

Reply:
why not?

Reply:
If an asset (producer) is turned off at a specific timestep in stage 1, in stage 2 the producer might have to be turned on and thus the pipe should not be disconnected.

Reply:
similar as with the other comment, this might be possible, but very unlikely. This only happens if we are already pushing the upper bounds of the source sizes in stage 1. For now I would like to keep it this way to improve computational time